File: TDD/TDD Coding Practice String Transformer/Code/Python/instruction/original/presenter.py
from interaction import *
from validator import ValidatingResult, ParamValidatingRule, Validator

class Presenter(object):
    def __init__(self, view, businesslogic):
        self.view = view
        self.businesslogic = businesslogic
        self.view.set_presenter(self)

        self.avail_selected_index = 0
        self.chain_selected_index = NONE_SELECTED_INDEX
        self.avail_transes = None
        self.chain_transes = []
        self.result_str = None

    def init(self):
        self.avail_transes = self.businesslogic.get_all_transes()
        self.view.on_init({AVAIL_TRANSES:self.avail_transes,
                           AVAIL_SELECTED_INDEX:self.avail_selected_index})

    def add_trans(self):
        avail_selected_trans = self.view.collect_add_trans_data()[AVAIL_SELECTED_TRANS]

        validating_result = self.validate(self.build_param_validating_rules_for_add(avail_selected_trans))
        if validating_result.is_succeeded:
            self.chain_transes.append(avail_selected_trans)

        self.update_chain_selected_index_for_add(avail_selected_trans, validating_result.failed_reason)
        self.update_avail_selected_index_for_add(avail_selected_trans, validating_result.failed_reason)

        self.view.on_add_trans({CHAIN_TRANSES:self.chain_transes,
                                CHAIN_SELECTED_INDEX:self.chain_selected_index,
                                AVAIL_SELECTED_INDEX:self.avail_selected_index})

    def build_param_validating_rules_for_add(self, avail_selected_trans):
        return [ParamValidatingRule(avail_selected_trans,
                                    Presenter.trans_not_specified,
                                    ValidatingResult.VRFR_AVAIL_TRANS_NOT_SPECIFIED),
                ParamValidatingRule(avail_selected_trans,
                                    self.already_existed_in_chain,
                                    ValidatingResult.VRFR_ADD_ALREADY_EXISTED_IN_CHAIN_TRANS)]

    def update_chain_selected_index_for_add(self, avail_selected_trans, validating_result_failed_reason):
        if validating_result_failed_reason == ValidatingResult.VRFR_AVAIL_TRANS_NOT_SPECIFIED:
            return

        self.chain_selected_index = self.chain_transes.index(avail_selected_trans)

    def update_avail_selected_index_for_add(self, avail_selected_trans, validating_result_failed_reason):
        if validating_result_failed_reason == ValidatingResult.VRFR_AVAIL_TRANS_NOT_SPECIFIED:
            self.avail_selected_index = 0
            return

        selected_index = self.avail_transes.index(avail_selected_trans)
        if validating_result_failed_reason == ValidatingResult.VRFR_ADD_ALREADY_EXISTED_IN_CHAIN_TRANS:
            self.avail_selected_index = selected_index
        else:
            self.avail_selected_index = \
                0 if Presenter.is_last_index(selected_index, self.avail_transes) else selected_index + 1

    def remove_trans(self):
        remove_trans_data = self.view.collect_remove_trans_data()
        chain_selected_trans = remove_trans_data[CHAIN_SELECTED_TRANS] if remove_trans_data else None

        validating_result = self.validate(self.build_param_validating_rules_for_remove(chain_selected_trans))

        self.update_chain_selected_index_for_remove(chain_selected_trans, validating_result.failed_reason)
        if validating_result.is_succeeded:
            self.chain_transes.remove(chain_selected_trans)
        self.update_avail_selected_index_for_remove(chain_selected_trans, validating_result.failed_reason)

        self.view.on_remove_trans({CHAIN_TRANSES:self.chain_transes,
                                   AVAIL_SELECTED_INDEX:self.avail_selected_index,
                                   CHAIN_SELECTED_INDEX:self.chain_selected_index})

    def build_param_validating_rules_for_remove(self, chain_selected_trans):
        return [ParamValidatingRule(self.chain_transes,
                                    Presenter.empty_list,
                                    ValidatingResult.VRFR_CHAIN_EMPTY),
                ParamValidatingRule(chain_selected_trans,
                                    Presenter.trans_not_specified,
                                    ValidatingResult.VRFR_CHAIN_TRANS_NOT_SPECIFIED)]

    def update_chain_selected_index_for_remove(self, chain_selected_trans, validating_result_failed_reason):
        if validating_result_failed_reason == ValidatingResult.VRFR_CHAIN_EMPTY:
            self.chain_selected_index = NONE_SELECTED_INDEX
            return

        if validating_result_failed_reason == ValidatingResult.VRFR_CHAIN_TRANS_NOT_SPECIFIED:
            self.chain_selected_index = 0
            return

        if len(self.chain_transes) == 1:
            self.chain_selected_index = NONE_SELECTED_INDEX
            return

        selected_index = self.chain_transes.index(chain_selected_trans)
        self.chain_selected_index = \
            0 if Presenter.is_last_index(selected_index, self.chain_transes) else selected_index

    def update_avail_selected_index_for_remove(self, chain_selected_trans, validating_result_failed_reason):
        if validating_result_failed_reason == ValidatingResult.VRFR_CHAIN_EMPTY:
            self.avail_selected_index = 0
            return

        if validating_result_failed_reason == ValidatingResult.VRFR_CHAIN_TRANS_NOT_SPECIFIED:
            return

        self.avail_selected_index = self.avail_transes.index(chain_selected_trans)

    def remove_all_transes(self):
        validating_result = self.validate(self.build_param_validating_rules_for_remove_all())
        if validating_result.is_succeeded:
            del self.chain_transes[:]
        self.update_chain_selected_index_for_remove_all()
        self.update_avail_selected_index_for_remove_all(validating_result.failed_reason)

        self.view.on_remove_all_transes({CHAIN_TRANSES:self.chain_transes,
                                         CHAIN_SELECTED_INDEX:self.chain_selected_index,
                                         AVAIL_SELECTED_INDEX:self.avail_selected_index})

    def build_param_validating_rules_for_remove_all(self):
        return [ParamValidatingRule(self.chain_transes,
                                    Presenter.empty_list,
                                    ValidatingResult.VRFR_CHAIN_EMPTY)]

    def update_chain_selected_index_for_remove_all(self):
        self.chain_selected_index = NONE_SELECTED_INDEX

    def update_avail_selected_index_for_remove_all(self, validating_result_failed_reason):
        if validating_result_failed_reason == ValidatingResult.VRFR_CHAIN_EMPTY:
            return

        self.avail_selected_index = 0

    def apply_trans_chain(self):
        source_str = self.view.collect_apply_trans_chain_data()[SOURCE_STR]

        self.result_str = self.businesslogic.transform(source_str, self.chain_transes)

        self.view.on_apply_trans_chain({RESULT_STR:self.result_str})

    def validate(self, param_validating_rules):
        validating_result = Validator.validate(param_validating_rules)
        if not validating_result.is_succeeded:
            self.view.on_validating_failed(validating_result.failed_reason)
        return validating_result

    def already_existed_in_chain(self, trans):
        return trans in self.chain_transes

    @staticmethod
    def is_last_index(index, lst):
        return index == len(lst) - 1

    @staticmethod
    def trans_not_specified(trans):
        return trans == None

    @staticmethod
    def empty_list(lst):
        return lst == []

File: TDD/TDD Coding Practice String Transformer/Code/Java/src/fayelab/tdd/stringtransformer/instruction/addall/Presenter.java
package fayelab.tdd.stringtransformer.instruction.addall;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.BiConsumer;
import java.util.function.Function;
import java.util.function.Supplier;

import fayelab.tdd.stringtransformer.instruction.addall.Entry.Key;
import fayelab.tdd.stringtransformer.instruction.addall.ValidatingResult.FailedReason;

import static java.util.Arrays.asList;
import static fayelab.tdd.stringtransformer.instruction.addall.Interaction.*;
import static fayelab.tdd.stringtransformer.instruction.addall.Entry.*;
import static fayelab.tdd.stringtransformer.instruction.addall.Entry.Key.*;
import static fayelab.tdd.stringtransformer.instruction.addall.ParamValidatingRule.*;
import static fayelab.tdd.stringtransformer.instruction.addall.OperData.*;

public class Presenter
{
    private View view;
    private BusinessLogic businessLogic;

    private int availSelectedIndex = 0;
    private int chainSelectedIndex = NONE_SELECTED_INDEX;
    private List<String> availTranses;
    private List<String> chainTranses = new ArrayList<>();
    private String resultStr;

    public Presenter(View view, BusinessLogic businessLogic)
    {
        this.view = view;
        this.businessLogic = businessLogic;
        view.setPresenter(this);
    }

    public void init()
    {
        availTranses = businessLogic.getAllTranses();
        view.onInit(interactionData(
                entry(AVAIL_TRANSES, availTranses),
                entry(AVAIL_SELECTED_INDEX, availSelectedIndex)));
    }

    public void addTrans()
    {
        operTrans(operData(Optional.of(this::collectViewDataForAdd), 
                           Optional.of(this::buildParamValidatingRulesForAdd), 
                           this::updatePresenterDataForAdd, 
                           this::presentViewDataForAdd));
    }

    private Optional<String> collectViewDataForAdd()
    {
        return Optional.ofNullable(view.collectAddTransData().get(AVAIL_SELECTED_TRANS).toStr());
    }

    private List<ParamValidatingRule<?>> buildParamValidatingRulesForAdd(Optional<String> viewData)
    {
        String availSelectedTrans = viewData.orElse(NONE_SELECTED_TRANS);
        return asList(
                paramValidatingRule(availSelectedTrans, Presenter::transNotSpecified, 
                        FailedReason.AVAIL_TRANS_NOT_SPECIFIED),
                paramValidatingRule(availSelectedTrans, this::alreadyExistedInChain, 
                        FailedReason.ADD_ALREADY_EXISTED_IN_CHAIN_TRANS));
    }

    private void updatePresenterDataForAdd(Optional<String> viewData, Optional<ValidatingResult> opValidatingResult)
    {
        String availSelectedTrans = viewData.orElse(NONE_SELECTED_TRANS);
        ValidatingResult validatingResult = opValidatingResult.get();
        if(validatingResult.isSucceeded())
        {
            chainTranses.add(availSelectedTrans);
        }
        updateChainSelectedIndexForAdd(availSelectedTrans, validatingResult.getFailedReason());
        updateAvailSelectedIndexForAdd(availSelectedTrans, validatingResult.getFailedReason());
    }

    private void updateChainSelectedIndexForAdd(String availSelectedTrans, FailedReason failedReason)
    {
        if(failedReason == FailedReason.AVAIL_TRANS_NOT_SPECIFIED)
        {
            return;
        }

        chainSelectedIndex = chainTranses.indexOf(availSelectedTrans);
    }

    private void updateAvailSelectedIndexForAdd(String availSelectedTrans, FailedReason failedReason)
    {
        if(failedReason == FailedReason.AVAIL_TRANS_NOT_SPECIFIED)
        {
            availSelectedIndex = 0;
            return;
        }

        int selectedIndex = availTranses.indexOf(availSelectedTrans);
        if(failedReason == FailedReason.ADD_ALREADY_EXISTED_IN_CHAIN_TRANS)
        {
            availSelectedIndex = selectedIndex;
        }
        else
        {
            availSelectedIndex = isLastIndex(selectedIndex, availTranses) ? 0 : selectedIndex + 1;
        }
    }

    private void presentViewDataForAdd()
    {
        view.onAddTrans(buildPresentViewDataForBuildingTransChain());
    }

    public void removeTrans()
    {
        operTrans(operData(Optional.of(this::collectViewDataForRemove), 
                           Optional.of(this::buildParamValidatingRulesForRemove), 
                           this::updatePresenterDataForRemove, 
                           this::presentViewDataForRemove));
    }

    private Optional<String> collectViewDataForRemove()
    {
        Map<Key, Value<?>> removeTransData = view.collectRemoveTransData();
        String chainSelectedTrans = removeTransData != null ? removeTransData.get(CHAIN_SELECTED_TRANS).toStr() : null;
        return Optional.ofNullable(chainSelectedTrans);
    }

    private List<ParamValidatingRule<?>> buildParamValidatingRulesForRemove(Optional<String> viewData)
    {
        String chainSelectedTrans = viewData.orElse(NONE_SELECTED_TRANS);
        return asList(
                paramValidatingRule(chainTranses, Presenter::emptyList, 
                        FailedReason.CHAIN_EMPTY),
                paramValidatingRule(chainSelectedTrans, Presenter::transNotSpecified, 
                        FailedReason.CHAIN_TRANS_NOT_SPECIFIED));
    }

    private void updatePresenterDataForRemove(Optional<String> viewData, Optional<ValidatingResult> opValidatingResult)
    {
        String chainSelectedTrans = viewData.orElse(NONE_SELECTED_TRANS);
        ValidatingResult validatingResult = opValidatingResult.get();
        updateChainSelectedIndexForRemove(chainSelectedTrans, validatingResult.getFailedReason());
        if(validatingResult.isSucceeded())
        {
            chainTranses.remove(chainSelectedTrans);
        }
        updateAvailSelectedIndexForRemove(chainSelectedTrans, validatingResult.getFailedReason());
    }

    private void updateChainSelectedIndexForRemove(String chainSelectedTrans, FailedReason failedReason)
    {
        if(failedReason == FailedReason.CHAIN_EMPTY)
        {
            chainSelectedIndex = NONE_SELECTED_INDEX;
            return;
        }

        if(failedReason == FailedReason.CHAIN_TRANS_NOT_SPECIFIED)
        {
            chainSelectedIndex = 0;
            return;
        }

        if(chainTranses.size() == 1)
        {
            chainSelectedIndex = NONE_SELECTED_INDEX;
            return;
        }

        int selectedIndex = chainTranses.indexOf(chainSelectedTrans);
        chainSelectedIndex = isLastIndex(selectedIndex, chainTranses) ? 0 : selectedIndex;
    }

    private void updateAvailSelectedIndexForRemove(String chainSelectedTrans, FailedReason failedReason)
    {
        if(failedReason == FailedReason.CHAIN_EMPTY)
        {
            availSelectedIndex = 0;
            return;
        }

        if(failedReason == FailedReason.CHAIN_TRANS_NOT_SPECIFIED)
        {
            return;
        }

        availSelectedIndex = availTranses.indexOf(chainSelectedTrans);
    }

    private void presentViewDataForRemove()
    {
        view.onRemoveTrans(buildPresentViewDataForBuildingTransChain());
    }

    public void removeAllTranses()
    {
        operTrans(operData(Optional.empty(), 
                           Optional.of(this::buildParamValidatingRulesForRemoveAll), 
                           this::updatePresenterDataForRemoveAll, 
                           this::presentViewDataForRemoveAll));
    }

    private List<ParamValidatingRule<?>> buildParamValidatingRulesForRemoveAll(Optional<?> emptyViewData)
    {
        return asList(paramValidatingRule(chainTranses, Presenter::emptyList, FailedReason.CHAIN_EMPTY));
    }

    private void updatePresenterDataForRemoveAll(Optional<?> emptyViewData, Optional<ValidatingResult> opValidatingResult)
    {
        ValidatingResult validatingResult = opValidatingResult.get();
        if(validatingResult.isSucceeded())
        {
            chainTranses.clear();
        }
        updateChainSelectedIndexForRemoveAll();
        updateAvailSelectedIndexForRemoveAll(validatingResult.getFailedReason());
    }

    private void updateChainSelectedIndexForRemoveAll()
    {
        chainSelectedIndex = NONE_SELECTED_INDEX;
    }

    private void updateAvailSelectedIndexForRemoveAll(FailedReason failedReason)
    {
        if(failedReason == FailedReason.CHAIN_EMPTY)
        {
            return;
        }

        availSelectedIndex = 0;
    }

    private void presentViewDataForRemoveAll()
    {
        view.onRemoveAllTranses(buildPresentViewDataForBuildingTransChain());
    }

    public void applyTransChain()
    {
        operTrans(operData(Optional.of(this::collectViewDataForApply), 
                           Optional.of(this::buildParamValidatingRulesForApply), 
                           this::updatePresenterDataForApply, 
                           this::presentViewDataForApply));
    }

    private Optional<String> collectViewDataForApply()
    {
        return Optional.ofNullable(view.collectApplyTransChainData().get(SOURCE_STR).toStr());
    }

    private List<ParamValidatingRule<?>> buildParamValidatingRulesForApply(Optional<String> viewData)
    {
        String sourceStr = viewData.get();
        return asList(
                paramValidatingRule(sourceStr, Presenter::emptyStr, 
                        FailedReason.SOURCE_STR_EMPTY),
                paramValidatingRule(sourceStr, Presenter::illegalSourceStr, 
                        FailedReason.SOURCE_STR_ILLEGAL),
                paramValidatingRule(chainTranses, Presenter::emptyList, 
                        FailedReason.CHAIN_EMPTY));
    }

    private void updatePresenterDataForApply(Optional<String> viewData, Optional<ValidatingResult> opValidatingResult)
    {
        ValidatingResult validatingResult = opValidatingResult.get();
        resultStr = validatingResult.isSucceeded() ? businessLogic.transform(viewData.get(), chainTranses) : "";
        updateAvailSelectedIndexForApply(validatingResult.getFailedReason());
    }

    private void updateAvailSelectedIndexForApply(FailedReason failedReason)
    {
        if(failedReason == FailedReason.CHAIN_EMPTY)
        {
            availSelectedIndex = 0;
        }
    }

    private void presentViewDataForApply()
    {
        view.onApplyTransChain(interactionData(
                entry(RESULT_STR, resultStr),
                entry(AVAIL_SELECTED_INDEX, availSelectedIndex)));
    }

    public void addAllTranses()
    {
        operTrans(operData(Optional.empty(), 
                           Optional.empty(), 
                           this::updatePresenterDataForAddAll, 
                           this::presentViewDataForAddAll));
    }

    private void updatePresenterDataForAddAll(Optional<?> emptyViewData, Optional<?> emptyValidatingResult)
    {
        chainTranses.clear();
        chainTranses.addAll(availTranses);
        updateChainSelectedIndexForAddAll();
        updateAvailSelectedIndexForAddAll();
    }

    private void updateChainSelectedIndexForAddAll()
    {
        chainSelectedIndex = chainTranses.size() - 1;
    }

    private void updateAvailSelectedIndexForAddAll()
    {
        availSelectedIndex = 0;
    }

    private void presentViewDataForAddAll()
    {
        view.onAddAllTranses(buildPresentViewDataForBuildingTransChain());
    }

    private <T> void operTrans(OperData<T> operData)
    {
        Optional<T> viewData = operData.getCollectViewDataFunc().orElse(() -> Optional.empty()).get();
        Optional<ValidatingResult> validatingResult = 
                operData.getBuildParamValidatingRulesFunc()
                        .flatMap(func -> Optional.of(validate(func.apply(viewData))));
        operData.getUpdatePresenterDataFunc().accept(viewData, validatingResult);
        operData.getPresentViewDataFunc().run();
    }

    private Map<Key, Value<?>> buildPresentViewDataForBuildingTransChain()
    {
        return interactionData(
                entry(CHAIN_TRANSES, chainTranses),
                entry(CHAIN_SELECTED_INDEX, chainSelectedIndex),
                entry(AVAIL_SELECTED_INDEX, availSelectedIndex));
    }

    private ValidatingResult validate(List<ParamValidatingRule<?>> rules)
    {
        ValidatingResult validatingResult = Validator.validate(rules);
        if(!validatingResult.isSucceeded())
        {
            view.onValidatingFailed(interactionData(
                    entry(VALIDATING_FAILED_REASON, validatingResult.getFailedReason())));
        }
        return validatingResult;
    }

    private boolean alreadyExistedInChain(String trans)
    {
        return chainTranses.contains(trans);
    }

    private static boolean isLastIndex(int index, List<?> list)
    {
        return index == list.size() - 1;
    }

    private static boolean transNotSpecified(String trans)
    {
        return trans == NONE_SELECTED_TRANS;
    }

    private static boolean emptyList(List<?> list)
    {
        return list.isEmpty();
    }

    private static boolean emptyStr(String str)
    {
        return str.isEmpty();
    }

    private static boolean illegalSourceStr(String str)
    {
        return str.matches(".*[\u4e00-\u9fa5]+.*");
    }
}

class OperData<T>
{
    private Optional<Supplier<Optional<T>>> collectViewDataFunc;
    private Optional<Function<Optional<T>, List<ParamValidatingRule<?>>>> buildParamValidatingRulesFunc;
    private BiConsumer<Optional<T>, Optional<ValidatingResult>> updatePresenterDataFunc;
    private Runnable presentViewDataFunc;

    static <T> OperData<T> operData(
            Optional<Supplier<Optional<T>>> collectViewDataFunc, 
            Optional<Function<Optional<T>, List<ParamValidatingRule<?>>>> buildParamValidatingRulesFunc, 
            BiConsumer<Optional<T>, Optional<ValidatingResult>> updatePresenterDataFunc, 
            Runnable presentViewDataFunc)
    {
        return new OperData<>(collectViewDataFunc, buildParamValidatingRulesFunc, 
                updatePresenterDataFunc, presentViewDataFunc);
    }

    private OperData(Optional<Supplier<Optional<T>>> collectViewDataFunc, 
                     Optional<Function<Optional<T>, List<ParamValidatingRule<?>>>> buildParamValidatingRulesFunc, 
                     BiConsumer<Optional<T>, Optional<ValidatingResult>> updatePresenterDataFunc, 
                     Runnable presentViewDataFunc)
    {
        this.collectViewDataFunc = collectViewDataFunc;
        this.buildParamValidatingRulesFunc = buildParamValidatingRulesFunc;
        this.updatePresenterDataFunc = updatePresenterDataFunc;
        this.presentViewDataFunc = presentViewDataFunc;
    }

    public Optional<Supplier<Optional<T>>> getCollectViewDataFunc()
    {
        return collectViewDataFunc;
    }

    public Optional<Function<Optional<T>, List<ParamValidatingRule<?>>>> getBuildParamValidatingRulesFunc()
    {
        return buildParamValidatingRulesFunc;
    }

    public BiConsumer<Optional<T>, Optional<ValidatingResult>> getUpdatePresenterDataFunc()
    {
        return updatePresenterDataFunc;
    }

    public Runnable getPresentViewDataFunc()
    {
        return presentViewDataFunc;
    }
}